core/src/apps/cardano/helpers/tx_builder.py

```diff
@@ -0,0 +1,188 @@
+        self.cbor_hash_builder.finish_collection()
+
+    def add_asset_group(self, policy_id: bytes, tokens_count: int) -> None:
+        self.state_machine.transition(TxBuilderStateMachine.ADD_ASSET_GROUP)
+        self.cbor_hash_builder.start_collection_at_key(
+            key=policy_id,
+            collection=LazyCollection(DICT, tokens_count),
+        )
+
+    def finish_asset_groups(self) -> None:
+        self.state_machine.transition(TxBuilderStateMachine.FINISH_ASSET_GROUPS)
+        self.cbor_hash_builder.finish_collection()
+
+    def add_token(self, asset_name_bytes: bytes, amount: int) -> None:
+        self.state_machine.transition(TxBuilderStateMachine.ADD_TOKEN)
+        self.cbor_hash_builder.add_item((asset_name_bytes, amount))
+
+    def finish_tokens(self) -> None:
+        self.state_machine.transition(TxBuilderStateMachine.FINISH_TOKENS)
+        self.cbor_hash_builder.finish_collection()
+
+    def add_fee(self, fee: int) -> None:
+        self.state_machine.transition(TxBuilderStateMachine.ADD_FEE)
+        self.cbor_hash_builder.add_item((TX_BODY_FEE_KEY, fee))
+
+    def add_ttl(self, ttl: int | None) -> None:
+        if ttl is not None:
+            self.state_machine.transition(TxBuilderStateMachine.ADD_TTL)
+            self.cbor_hash_builder.add_item((TX_BODY_TTL_KEY, ttl))
+
+    def start_certificates(self, certificates_count: int) -> None:
+        self.state_machine.transition(TxBuilderStateMachine.START_CERTIFICATES)
+        self.cbor_hash_builder.start_collection_at_key(
+            key=TX_BODY_CERTIFICATES_KEY,
+            collection=LazyCollection(LIST, certificates_count),
+        )
+
+    def add_certificate(
+        self,
+        cborized_certificate: cbor.CborSequence,
+        owners_and_relays_count: Tuple[int, int] | None = None,
+    ) -> None:
+        if owners_and_relays_count is None:
+            self.state_machine.transition(TxBuilderStateMachine.ADD_SIMPLE_CERTIFICATE)
+            self.cbor_hash_builder.add_item(cborized_certificate)
+        else:
+            self.state_machine.transition(
+                TxBuilderStateMachine.ADD_POOL_REGISTRATION_CERTIFICATE
+            )
+            owners_count, relays_count = owners_and_relays_count
+            owners = LazyCollection(LIST, owners_count)
+            relays = LazyCollection(LIST, relays_count)
+            assert (
+                len(cborized_certificate) == 10
+                and cborized_certificate[7] is None
+                and cborized_certificate[8] is None
+            )
+            cborized_certificate = list(cborized_certificate)
+            cborized_certificate[7] = owners
+            cborized_certificate[8] = relays
+            self.cbor_hash_builder.add_item(cborized_certificate)
+            self.cbor_hash_builder.active_items_stack += [relays, owners]
+
+    def start_pool_owners(self) -> None:
+        pass  # The header of the owner list was already serialized when the certificate was added
```

Comment:
Sadly I don't understand how or why it was already serialized 😄  I'm trying to think of a way to simplify the pool registration case, but I can't come up with anything atm. I feel like the `add_certificate` function knows too much details about the structure. I'll take a look at this again later (or we can do that together).